plugins/algoltek-usbcr/fu-algoltek-usbcr-device.c

```diff
@@ -0,0 +1,671 @@
+					       st->data,
+					       st->len,
+					       buf,
+					       bufsz,
+					       error);
+}
+
+static gboolean
+fu_algoltek_usbcr_device_do_read_spi(FuAlgoltekUsbcrDevice *self,
+				     guint16 addr,
+				     guint8 *buf,
+				     guint8 bufsz,
+				     GError **error)
+{
+	g_autoptr(GByteArray) st = fu_struct_ag_usbcr_spi_cdb_new();
+
+	fu_struct_ag_usbcr_spi_cdb_set_cmd(st, FU_AG_USBCR_SCSIOP_VENDOR_EEPROM_RD);
+	fu_struct_ag_usbcr_spi_cdb_set_addr(st, addr);
+	fu_struct_ag_usbcr_spi_cdb_set_bufsz(st, bufsz);
+	fu_struct_ag_usbcr_spi_cdb_set_tag(st, FU_AG_SPECIFY_EEPROM_TYPE_TAG);
+	fu_struct_ag_usbcr_spi_cdb_set_valid(st, FU_AG_SPIFLASH_VALID);
+
+	return fu_block_device_sg_io_cmd_read(FU_BLOCK_DEVICE(self),
+					      st->data,
+					      st->len,
+					      buf,
+					      bufsz,
+					      error);
+}
+
+static gboolean
+fu_algoltek_usbcr_device_command_wren(FuAlgoltekUsbcrDevice *self, GError **error)
+{
+	guint8 buf[1] = {0};
+	FuAgUsbcrRegSetup regs[] = {
+	    {0xC8, 0x04},
+	    {0xCA, 0x01},
+	    {0x400, FU_AG_USBCR_WREN},
+	    {0xC9, 0x01},
+	    {0xC8, 0x05},
+	};
+
+	for (guint i = 0; i < G_N_ELEMENTS(regs); i++) {
+		if (!fu_algoltek_usbcr_device_write_reg(self,
+							regs[i].reg,
+							regs[i].val,
+							FU_AG_USBCR_RD_WR_XDATA,
+							error))
+			return FALSE;
+	}
+	do {
+		if (!fu_algoltek_usbcr_device_read_reg(self,
+						       0xC8,
+						       buf,
+						       sizeof(buf),
+						       FU_AG_USBCR_RD_WR_XDATA,
+						       error))
+			return FALSE;
+	} while (buf[0] & 0x01);
+
+	/* success*/
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usbcr_device_command_wrsr(FuAlgoltekUsbcrDevice *self, gboolean en, GError **error)
+{
+	guint8 buf[1] = {0};
+	FuAgUsbcrRegSetup regs[] = {
+	    {0xC8, 0x04},
+	    {0xCA, 0x01},
+	    {0x400, FU_AG_USBCR_WRSR},
+	    {0x401, 0x00},
+	    {0xC9, 0x02},
+	    {0xC8, 0x05},
+	};
+	if (en == FU_AG_BLOCK_MODE_EN)
+		regs[3].val = 0x0C;
+	for (guint i = 0; i < G_N_ELEMENTS(regs); i++) {
+		if (!fu_algoltek_usbcr_device_write_reg(self,
+							regs[i].reg,
+							regs[i].val,
+							FU_AG_USBCR_RD_WR_XDATA,
+							error))
+			return FALSE;
+	}
+	do {
+		if (!fu_algoltek_usbcr_device_read_reg(self,
+						       0xC8,
+						       buf,
+						       sizeof(buf),
+						       FU_AG_USBCR_RD_WR_XDATA,
+						       error))
+			return FALSE;
+	} while (buf[0] & 0x01);
+
+	/* success*/
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usbcr_device_command_rdsr(FuAlgoltekUsbcrDevice *self, GError **error)
+{
+	guint8 buf[1] = {0};
+	FuAgUsbcrRegSetup regs[] = {
+	    {0xC8, 0x04},
+	    {0xCA, 0x01},
+	    {0x400, FU_AG_USBCR_RDSR},
+	    {0xC9, 0x01},
+	    {0xC4, 0x01},
+	    {0xC7, 0x00},
+	    {0xC8, 0x07},
+	};
+	for (guint i = 0; i < G_N_ELEMENTS(regs); i++) {
+		if (!fu_algoltek_usbcr_device_write_reg(self,
+							regs[i].reg,
+							regs[i].val,
+							FU_AG_USBCR_RD_WR_XDATA,
+							error))
+			return FALSE;
+	}
+	do {
```

Comment:
i think this needs to use `fu_device_retry` too

Reply:
Updated, use fu_device_retry_full() instead of do-while.